server/indexer.go

```diff
@@ -0,0 +1,203 @@
+package server
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authante "github.com/cosmos/cosmos-sdk/x/auth/ante"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/evmos/ethermint/rpc/backend"
+	rpctypes "github.com/evmos/ethermint/rpc/types"
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/libs/log"
+	tmtypes "github.com/tendermint/tendermint/types"
+	dbm "github.com/tendermint/tm-db"
+
+	ethermint "github.com/evmos/ethermint/types"
+	evmtypes "github.com/evmos/ethermint/x/evm/types"
+)
+
+var (
+	_ ethermint.EVMTxIndexer = &KVIndexer{}
+
+	TxIndexKeyLength = len(TxIndexKey(0, 0))
+)
+
+const (
+	KeyPrefixTxHash  = 1
+	KeyPrefixTxIndex = 2
+)
+
+func TxHashKey(hash common.Hash) []byte {
+	return append([]byte{KeyPrefixTxHash}, hash.Bytes()...)
+}
+
+func TxIndexKey(blockNumber int64, txIndex int32) []byte {
+	bz1 := sdk.Uint64ToBigEndian(uint64(blockNumber))
+	bz2 := sdk.Uint64ToBigEndian(uint64(txIndex))
+	return append(append([]byte{KeyPrefixTxIndex}, bz1...), bz2...)
+}
+
+func parseBlockNumberFromKey(key []byte) (int64, error) {
+	if len(key) != TxIndexKeyLength {
+		return 0, fmt.Errorf("wrong tx index key length, expect: %d, got: %d", TxIndexKeyLength, len(key))
+	}
+
+	return int64(sdk.BigEndianToUint64(key[1:9])), nil
+}
+
+// LoadLastBlock returns -1 if db is empty
+func LoadLastBlock(db dbm.DB) (int64, error) {
+	it, err := db.ReverseIterator([]byte{KeyPrefixTxIndex}, []byte{KeyPrefixTxIndex + 1})
+	if err != nil {
+		return 0, err
+	}
+	defer it.Close()
+	if !it.Valid() {
+		return -1, nil
+	}
+	return parseBlockNumberFromKey(it.Key())
+}
+
+// LoadFirstBlock returns -1 if db is empty
+func LoadFirstBlock(db dbm.DB) (int64, error) {
+	it, err := db.Iterator([]byte{KeyPrefixTxIndex}, []byte{KeyPrefixTxIndex + 1})
+	if err != nil {
+		return 0, err
+	}
+	defer it.Close()
+	if !it.Valid() {
+		return -1, nil
+	}
+	return parseBlockNumberFromKey(it.Key())
+}
+
+type KVIndexer struct {
```

Comment:
should the indexer have a shutdown "hook", so that it gracefully closes the DB on the process exit?

Reply:
or in `EVMIndexerService`?

Reply:
It seems to be a common practice to not close the DB, for example, the [app db in cosmos-sdk](https://github.com/cosmos/cosmos-sdk/blob/main/baseapp/baseapp.go#L49), and tx indexer DB in tendermint.